### PR TITLE
Hanami::Interactor Only symbolize keys of result payload

### PR DESCRIPTION
**Pull request by alexd16**
Fixes: https://github.com/hanami/utils/issues/159

/cc @jodosha

### CODE REVIEW

File: lib/hanami/utils/hash.rb [hunk 1]
```diff
@@ -62,11 +62,10 @@ def initialize(hash = {}, &blk)
       #
       #   hash.keys    # => [:a, :b]
       #   hash.inspect # => {:a=>23, :b=>{:c=>["x", "y", "z"]}}
-      def symbolize!
+      def symbolize!(deep = true)
```
Comment (alexd16) [inline lib/hanami/utils/hash.rb]: Yeah, that makes sense. Will update the PR.

File: test/hash_test.rb [hunk 2]
```diff
@@ -33,29 +33,40 @@
...
-      nested = Hanami::Utils::Hash.new('metadata' => WrappingHash.new('coverage' => 100))
-      nested.symbolize!
+    describe 'with deep false' do
+      it 'does not simbolize nested hashes' do
+        hash = Hanami::Utils::Hash.new('nested' => { 'key' => 'value' })
+        hash.symbolize!(deep: false)
 
-      nested[:metadata].must_be_kind_of Hanami::Utils::Hash
-      nested[:metadata][:coverage].must_equal(100)
+        hash[:nested]['key'].must_equal('value')
```
Comment (joneslee85) [inline test/hash_test.rb]: What do you think about checking the `keys` of `hash[:nested]` instead? To me it is clearer in intent as we do only assert that the key is not symoblized

File: test/interactor_test.rb [hunk 3]
```diff
@@ -412,6 +412,11 @@ def initialize(_user, params)
       result.b.must_equal 2
     end
 
+    it 'returns hash values passed in the payload' do
```
Comment (joneslee85) [inline test/interactor_test.rb]: What do you think about changing the description to `it "returns a hash value that is symbolized shallowly"`?

File: lib/hanami/utils/hash.rb [hunk 4]
```diff
@@ -48,7 +48,7 @@ def initialize(hash = {}, &blk)
         @hash.default_proc = blk
       end
 
-      # Convert in-place all the keys to Symbol instances, nested hashes are converted too.
+      # Convert in-place all the keys to Symbol instances. Nested hashes are converted too if deep argument is truthy.
```
Comment (joneslee85) [inline lib/hanami/utils/hash.rb]: What do you think about changing the wording to `, optional nested hashes`

Comment (alexd16) [inline lib/hanami/utils/hash.rb]: Like this: Convert in-place all the keys to Symbol instances, optionally nested hashes are converted too if deep argument is truthy. ?

File: test/interactor_test.rb (hunk 3 above)
Comment (alexd16) [inline test/interactor_test.rb]: But the value is not symbolized, I am not sure if that is very clear. This test is to verify that the values of the Hash passed as an argument are not changed.

For example: Currently if we have
result = Interactor::Result.new({ m: { "key" => "value" } })
result.m => { key: "value" }
and it should be
result.m => { "key" => "value" }

To be honest I am not sure if we need this additional test. We already have a test before this one that has:

```ruby
    it 'returns all the values passed in the payload' do
      result = Hanami::Interactor::Result.new(a: 1, b: 2)
      result.a.must_equal 1
      result.b.must_equal 2
    end
``` 

Given that description the current behaviour is already failing it. Our test code is just not verifying it.
Maybe we could change this test to contemplate more cases passing a more complex hash in the constructor.

What do you think?

File: test/hash_test.rb (hunk 2 above)
Comment (alexd16) [inline test/hash_test.rb]: Yeah that makes sense. Will update the PR, thank you

**Comment by jodosha** [comment]
@alexd16 Hi and thanks for this PR.

Can we separate `Utils::Hash#symbolize!` from `Utils::Hash#deep_symbolize!`? Thanks.

**Comment by alexd16** [comment]
@jodosha  Yeah sure, not problem. 
But that could break current usage of the method symbolize! right? Is that ok?
I could create a separate method shallow_symbolize! to avoid this. 

What do you prefer?

**Comment by jodosha** [comment]
@alexd16 

> But that could break the current usage of the method symbolize! if someone is using the method expecting that it does a deep symbolize right?

Yes it's true, but we're still under 1.0, this is the right time to do this change as we'll hit that milestone soon.

File: lib/hanami/utils/hash.rb (hunk 4 above)
Comment (joneslee85) [inline lib/hanami/utils/hash.rb]: yeah

**Comment by alexd16** [comment]
@jodosha Sorry I didn't pick up the PR for some time. 

I have another question about separating symbolize! method into two. 

The current implementation searches every value that responds to the to_hash method and wraps it into an instance of the Utils::Hash class. 

I think this was done only to be able to call the symbolize! method for those hashes, and since we are no longer doing that I was wondering if we still need to do it.

What do you think?

File: lib/hanami/utils/hash.rb [hunk 5]
```diff
@@ -61,11 +61,34 @@ def initialize(hash = {}, &blk)
...
+      def symbolize!
+        keys.each do |k|
+          v = delete(k)
+          self[k.to_sym] = v
+        end
+
+        self
+      end
+
+      # Convert in-place all the keys to Symbol instances, nested hashes are converted too.
```
Comment (cllns) [inline lib/hanami/utils/hash.rb]: What about nested arrays of hashes? Can you add a test for that please? 🌸

Comment (cllns) [inline lib/hanami/utils/hash.rb]: What about nested arrays of hashes? Can you add a test for that please? 🌸

Comment (alexd16) [inline lib/hanami/utils/hash.rb]: For the deep_symbolize! method I was keeping the behaviour that exists currently in master. From what I see it does not symbolize nested array of hashes, but I noticed that there is a PR to support that thought: https://github.com/hanami/utils/pull/161

Should I add that behaviour in this PR?

**Comment by AlfonsoUceda** [comment]
@alexd16 thanks for the PR, could you add a test if the new method symbolize deep hashed with 3 or 4 levels? thanks ;)

**Comment by AlfonsoUceda** [comment]
to me it's ok, I think it's ready to be merged cc @jodosha

**Comment by jodosha** [comment]
@alexd16 Thanks for this fix. LGTM 👍 

@alexd16 @AlfonsoUceda We should verify that all the other Hanami gems work fine with this change.

**Comment by alexd16** [comment]
@jodosha Happy to help. Glad I could have my first open-source contribution accepted 😄 

About verifying that this works with the other Hanami gems, how can I do that? Not sure how to proceed or If I need to do anything else.

**Comment by AlfonsoUceda** [comment]
@alexd16 you should check every hanami component using hanami/utils, and in the Gemfile point to hanami/utils against your fork and branch and then run the tests. If you have problems running tests you could create a PR and delegate the work to travis.

**Comment by mereghost** [comment]
And this will need to be on the upgrade notes as this breaks the behavior set in #116 since it split the methods.

**Comment by AlfonsoUceda** [comment]
@alexd16 could you update your fork with the last changes in upstream? your fork is 0.8.0 version and we are no in 0.9.1. Your fork needs to be updated for testing this branch.

Thanks!

**Comment by alexd16** [comment]
@AlfonsoUceda Should I update with master or 0.9.1?

**Comment by AlfonsoUceda** [comment]
@alexd16 with master I prefer so you'll have your fork up to date.

**Comment by alexd16** [comment]
@AlfonsoUceda Done.

Btw I ran the tests of the other Hanami gems locally and the hanami-model and hanami-controller have failing specs with this change.

I am just not sure how to fix those without this being merged.

**Comment by jodosha** [comment]
@alexd16 Thank you. I finally merged this. 👍

**Comment by oana-sipos** [comment]
Thanks @alexd16 for your contribution and for giving me the idea to write a small piece of documentation on how to proceed when you first contribute with code ;) very valid question! 👍

**Pull request merged**